apps/cfp/event_tickets_lottery.py

```diff
@@ -0,0 +1,126 @@
+from random import shuffle
+from sqlalchemy import func
+
+from flask import render_template, current_app as app
+from flask_mailman import EmailMessage
+
+from main import db
+from models.cfp import WorkshopProposal, Proposal
+from models.event_tickets import EventTicket
+from models.site_state import SiteState, refresh_states
+
+from ..common.email import from_email
+
+from . import cfp
+
+
+@cfp.cli.command("lottery")
+def lottery():
+    # In theory this can be extended to other types but currently only workshops & youthworkshops care
+    workshops = (
+        WorkshopProposal.query.filter_by(requires_ticket=True, type="workshop")
+        .filter(Proposal.state.in_(["accepted", "finalised"]))
+        .all()
+    )
+
+    app.logger.info(f"Running lottery for {len(workshops)} workshops")
+    winning_tickets = run_lottery(workshops)
+    app.logger.info(f"{len(winning_tickets)} won")
+
+    youthworkshops = (
+        WorkshopProposal.query.filter_by(requires_ticket=True, type="youthworkshop")
+        .filter(Proposal.state.in_(["accepted", "finalised"]))
+        .all()
+    )
+    app.logger.info(f"Running lottery for {len(youthworkshops)} youthworkshops")
+    winning_tickets = run_lottery(youthworkshops)
+    app.logger.info(f"{len(winning_tickets)} won")
+
+
+def run_lottery(ticketed_proposals):
+    """
+    Here are the rules for the lottery.
+    * Each user can only have one lottery ticket per workshop
+    * A user's lottery tickets are ranked by preference
+    * Drawings are done by rank
+    * Once a user wins a lottery their other tickets are cancelled
+    """
+    # Copy because we don't want to change the original
+    ticketed_proposals = ticketed_proposals.copy()
+    lottery_round = 0
+
+    winning_tickets = []
+
+    app.logger.info(f"Found {len(ticketed_proposals)} proposals to run a lottery for")
+    # Lock the lottery
+    signup = SiteState.query.get("signup_state")
+    if not signup:
+        raise Exception("'signup_state' not found.")
+
+    # This is the only state for running the lottery
+    signup.state = "run-lottery"
```

Comment:
I think this should check the state of the lottery before doing this, otherwise you risk re-entry.